<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Taking local control of genomics machines through BWA | Genomic Cybersecurity</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="Taking local control of genomics machines through BWA" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Summary" />
<meta property="og:description" content="Summary" />
<link rel="canonical" href="" />
<meta property="og:url" content="" />
<meta property="og:site_name" content="Genomic Cybersecurity" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2018-12-19T08:46:50-06:00" />
<meta property="article:modified_time" content="2019-08-12T14:31:28-6:00" />
<script type="application/ld+json">
{"description":"Summary","@type":"BlogPosting","url":"","headline":"Taking local control of genomics machines through BWA","dateModified":"2019-04-22T20:46:01+00:00","datePublished":"2019-04-22T20:46:01+00:00","mainEntityOfPage":{"@type":"WebPage","@id":""},"@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="assets/main.css"><link type="application/atom+xml" rel="alternate" href="feed.xml" title="Genomic Cybersecurity" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">Genomic Cybersecurity</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Taking local control of genomics machines through BWA</h1>
    <p class="post-meta">
      <time class="dt-published" datetime="2019-04-22T20:46:01+00:00" itemprop="datePublished">April 22, 2019
      </time></p>
      <time class="dt-modified" datetime="2019-08-12T12:19:30+00:00" itemprop="dateModified">August 12, 2019
      </time></p>

   </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h2 id="summary">Summary</h2>
<hr />

<p>In this post we discuss cybersecurity risks posed to systems engaged in genomic analysis, using a standard <strong>best-practices</strong> genomic pipeline. This post details the existence and consequences of an arbitrary code execution vulnerability affecting the most recent version of BWA on Github, including how it was discovered, exploit development, and demonstration of launching the exploit in a high-fidelity model environment.  Because of the standards of storage and processing in pipeline environments, the vulnerability can be exploited to access and modify patient data.  The vulnerable code can be reached and exploited by running <strong>“bwa mem”</strong> with a maliciously crafted .ALT file residing locally alongside other legitimate index files.  Therefore, the application must be run locally to reach exploitation, but the payload contained in a maliciously crafted .ALT file, may be delivered remotely.  The demonstration described in this post of an attack exploiting this vulnerability requires a presence on the target system’s network.</p>

<h2 id="implications">Implications</h2>
<hr />

<p>Because of the risks posed by the vulnerabilities described below, <em>medical/clinical genomic information</em>, <em>intellectual property related to genomic/genetic modification</em> and <em>forensic tests based on genomic matching</em> are potentially at risk if they are processed on servers doing genomic analysis.</p>

<h2 id="mitigations">Mitigations</h2>
<hr />

<p>In the section below we demonstrate how certain issues of cybersecurity hygiene, in combination with some long-term changes in the standards for hosting data can mitigate these vulnerabilities. In addition, we have submitted an <a href="https://github.com/lh3/bwa/pull/232">issue</a> and the maintainers have developed a <a href="https://github.com/lh3/bwa/commit/20d0a13092aa4cb73230492b05f9697d5ef0b88e">patch</a> to the public-facing BWA GitHub page: <a href="https://github.com/lh3/bwa">https://github.com/lh3/bwa</a></p>

<p>This issue has been reported to MITRE/US-CERT/NIST, as <a href="https://nvd.nist.gov/vuln/detail/CVE-2019-10269">CVE-2019-10269</a>.</p> This report characterizes the vulnerability as critical/high because of its potential to allow unauthorized disclosure of information, unauthorized modification and disruption of service.

<h2 id="background-on-next-gen-sequencing">Background on Next Gen Sequencing</h2>
<hr />

<p>The low cost of high throughput sequencing, primary by assembly of short genomic reads (known as Next Generation Sequencing or NGS), has fundamentally changed the scale and scope of genomic analysis.<sup><a href="#ref1">1</a></sup> The main idea behind NGS is to take gigabit scale genomic sequences and cut them into short strings (roughly 100-300 letters long) at high redundancy, with relatively low error rates and then use software to assemble the short strings (known as reads). While the fine-tuning of this process has dramatically reduced the start-to-finish cost of genomics, it has also induced heavy computational burdens. An average human genome may cost only a few hundred dollars to sequence (circa 2018) and assemble in software, but produces several terabytes of data in the process and requires machines with large amount of RAM to complete analysis. The rapid growth of this field has led to a surge of genomic tools and pipelines, many of which do nearly identical tasks and have extremely variable coding standards.<sup><a href="#ref2">2</a></sup></p>

<p><img src="assets/Whole_genome_shotgun_sequencing_versus_Hierarchical_shotgun_sequencing.png" alt="NGS" /><em>Visualization of whole genome sequencing, remixed with permission from Commins, J., Toft, C., Fares, M. A. “Computational Biology Methods and Their Application to the Comparative Genomics of Endocellular Symbiotic Bacteria of Insects.” Biol. Procedures Online (2009). Accessed via Springer Images. Original <a href="https://commons.wikimedia.org/wiki/File:Whole_genome_shotgun_sequencing_versus_Hierarchical_shotgun_sequencing.png">here</a>.</em></p>

<h2 id="best-practices-pipeline-for-variant-calling">Best Practices Pipeline for Variant Calling</h2>
<hr />

<p>One of the most important uses of whole genome sequencing is variant-calling. Variants can be used to directly observe genetic mutations. This can have clinical, scientific or forensic applications. In variant calling, the data typically arrive to the pipeline as raw reads from an NGS machine and leave the pipeline as a Variant Call Format (VCF) file, which lists each variable position relative to a reference genome.<sup><a href="#ref3">3</a></sup> The process of associating raw reads with a known reference genome, is referred to as <em>mapping</em>.</p>

<p>Different flavors of variant discovery require different steps, depending on if they pool samples, are germline (meaning fixed at birth) or somatic (meaning changing in the body) and whether they are single mutations or variations in copy number. In general, however there is a shared preprocessing set of steps. The best-practices listed below were developed by the BROAD Institute which sequences and assemble large numbers of genomes and has produced some of the most commonly used softwares for variant discovery.<sup><a href="#ref4">4</a></sup></p>

<p><img src="assets/dznasg7toiq1.png" alt="preprocess" /> <em>Original protocol</em> <a href="https://software.broadinstitute.org/gatk/best-practices/workflow?id=11165">here</a>.</p>

<p>Data comes into the preprocessing stages as <a href="http://support.illumina.com/content/dam/illumina-support/help/BaseSpaceHelp_v2/Content/Vault/Informatics/Sequencing_Analysis/BS/swSEQ_mBS_FASTQFiles.htm">FASTQ Files</a>. These files are then processed using a tool called <a href="https://github.com/lh3/bwa">BWA</a> that performs the task of <strong>“Map to Reference”</strong>. This produces a <a href="https://support.illumina.com/help/BS_App_RNASeq_Alignment_OLH_1000000006112/Content/Source/Informatics/BAM-Format.htm">BAM File</a>. Following this, duplicates are marked and a set of quality scores are estimated using a tool known as <a href="https://github.com/broadinstitute/picard">Picard</a>. The outcome of this is an <em>Analysis-Ready</em> BAM file.</p>

<p><img src="assets/tzm69d8e2spl.png" alt="somatic" /> This protocol describes the best practices for determining short germline variants - using <strong>BWA</strong>, followed by Picard, then followed by a number of functions within the GATK software. <em>Original posted</em> <a href="https://software.broadinstitute.org/gatk/best-practices/workflow?id=11145">here.</a></p>

<h2 id="the-vulnerability-and-exploit">The vulnerability and exploit</h2>
<hr />

<p>Given the above listed best-practices, this blog attempts to determine if it is possible to use an early step in a common software pipeline to exploit the outcome of a genomic test. Furthermore, we try to determine if we can change the raw data, limiting the ability of the user to detect this exploit after the fact. We do not fool ourselves into believing that the exploit we are attempting to craft is novel or particularly complex. The vulnerabilities that are included in this study are well-known, namely use of ARP poisoning to intercept and modify FTP traffic and <code class="highlighter-rouge">getc buffer overflow</code> vulnerabilities. In simulating our environment, we accept some security assumptions (namely, disabling of ASLR and stack canaries to make the exploit easier to visualize). Other assumptions are made from our experience of how these genomics pipelines are executed in-practice. Assuming .ALT files, and <em>in fact</em> many genomic databases, will be delivered over FTP is not only supported as a by-product of BWA’s associated bwa.kit software, but by the standard for holding and delivering genomic databases. The same is true for storing files as read/write-enabled and a lack of file integrity protections such as externally-stored file hashes.</p>

<h3 id="background">Background</h3>
<hr />

<p>In 2014, the developers of BWA added the ALT-aware capacity.<sup><a href="#ref5">5</a></sup> This capability allowed for a more statistically robust model for mapping genomic reads to a reference genome. Instead of assuming a canonical genome as a reference, the ALT-aware capability allows users to map reads to a population, which is effectively a distributional representation of the genome. This capacity created the need for a new file in BWA, the .ALT file. This file is constructed using a large collection of genomic variants (a challenge for individual users). And since the target of this file is constantly changing (as new individuals are added to the population of sequenced genomes), there is an advantage to storing these files at a central repository where anyone can access them. For the Human Genome, the US National Center for Biotechnology Information (NCBI) has provided resources for the storage and delivery of these files as a tarred and gzipped directory of indices <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/</a>. BWA provides a tool called bwa.kit which uses this repository to collect the most recent version of the file. BWA requires the program to access a set of index files, one of which, the .ALT file, stores the variable reference genome. When the user runs BWA in mapping mode (BWA MEM), these index files are read into memory first, then the read files are individually mapped to the reference genome. The result is a BAM file that stores the mapping of each read to the reference genome.</p>

<h3 id="steps-in-discovery">Steps in discovery</h3>
<hr />

<p>We discovered an exploitable arbitrary code execution vulnerability in the most recent version of BWA on GitHub while researching the security of BWA’s most common usage, BWA-MEM, in ALT-aware mode.</p>

<p>The vulnerable function allocates a 1024 byte buffer to store the contents of a .ALT file. Because the file size is not checked prior to allocating the buffer, a classic buffer overflow vulnerability exists. BWA must be run with the <strong>mem</strong> command line option to reach the vulnerable code.  Additionally, .ann, .amb, .pac, .bwt, .sa, and .ALT files must be present in order for execution to reach the vulnerable code.  These index files can be retrieved from running <strong>bwa index</strong> or delivered to the user by NCBI. The developers have acknowledged the <strong>bwa mem</strong> option to be the most commonly used BWA option by researchers. Though this vulnerability was only tested in BWA compiled for 32-bit and 64-bit x86 Linux, it is believed to exist in BWA compiled for any operating system. As indicated in the code comment in the figure below, BWA developers appear to be aware that a bug exists but may not be aware of its repercussions.</p>

<p><img src="assets/bwaVuln.png" alt="bwaVuln" /> <em>Function containing vulnerable code in BWA.</em></p>

<h3 id="crafting-an-exploit">Crafting an exploit</h3>
<hr />

<p>We make a few crucial security assumptions in crafting this exploit. To craft this exploit we required that ASLR and stack canaries are turned off. While this is not standard practice, it allows us to dramatically ease crafting the exploit for teaching purposes. The second assumption is that FASTQ files are read/write enabled. This <em>is</em> typical of normal environments, based on our own experience with how these files are stored on genomic servers.</p>

<p>As mentioned above, the bug can be encountered by using the <strong>bwa mem</strong> command in the same directory as .ann, .amb, .pac, .bwt, .sa files and a .ALT file with a line exceeding 1024 bytes in length. <strong>bwa mem</strong> uses several intermediate index files, the additional files that are required to be present, to allow mapping to the reference genome.  BWA takes in the prefix of the index files as well as two FASTQ files. We are running this analysis in paired-end mode, which is the current standard in genomic analysis.</p>

<p>After parsing the command line options to determine its branch of execution, BWA opens and reads each of the index files.  After reading the other index files, it attempts to open the .ALT file and store its contents into a locally created, static buffer. The offending function call responsible for writing the file contents into memory is <code class="highlighter-rouge">fgetc()</code>.  The function compares each character as its read in for <code class="highlighter-rouge">EOF, \t, \ n,</code> or <code class="highlighter-rouge">\r</code>, in which case it stops writing the current line’s contents to memory. If none of these characters are found it continues reading. Subsequently, providing a file named with the same prefix as the index files but with the .ALT extension that also contains a line made up of more than the allocated 1024 bytes and none of the aforementioned ASCII characters will overflow the buffer.</p>

<p>A user can create a segmentation fault if <strong>bwa mem</strong> is used with the index files and a .ALT file containing a line of 5200 bytes or more. This is shown in the figure below.</p>

<p><img src="assets/bwaSegFault.png" alt="bwaSegFault" /> <em><br /><br />Segmentation fault during execution due to size of file. Note: The file consists of only one line with 5200 characters.</em></p>

<p>The extra bytes are allocated to store other local variables including some of the index files’ names, a hash map, the incrementor variable indicating the current iteration’s buffer offset, and the .ALT file pointer.  The latter three variables are important to note because they will be crucial to the success of the exploit.  Running <strong>bwa mem</strong> with the large .ALT file in a debugger (we used GDB) reveals that the program crashes prior to returning from the vulnerable function, actually crashing during a call to <code class="highlighter-rouge">fgetc()</code>. Looking a little closer, GDB shows that the segmentation fault occurs due to an attempt to read from an invalid memory address (see the figure below).</p>

<p><img src="assets/firstSegFault64bit.png" alt="firstSegFault64bit" /> <em>Segmentation fault during execution due to reading from unmapped (invalid) memory address 0x555555767f stored in RDX+0x10.</em></p>

<p>This could be due to the file contents overwriting a local variable specifying where to read, such as the file pointer.  Examining memory contents of local variables reveals that the last character written has indeed overwritten the last byte of the file pointer.</p>

<p><img src="assets/filePointerChange.png" alt="filePointerChange" /> <em>First ‘A’ (0x41) byte being written over file pointer value.</em></p>

<p>This pointer to the stream must be kept intact to continue reading the file contents into memory.  Subsequently, the offset of this local variable from the start of the buffer must be calculated to determine where to write the pointer value.  Observing the value of RAX at the time the offending character is written gives the value of the offset stored in the local variable <em>i</em>.  In this case, writing the address value 0x1428 (5160) bytes into the payload will preserve the pointer.  Once the value of the pointer to the file object has been written in the appropriate place, the program continues past this point, but segmentation faults again without control of the instruction pointer.</p>

<p>Looking more closely at the other local variables in memory surrounding the file pointer reveals that the pointer to the hash map and the incrementor variable mentioned above are also being overwritten.  As a result, first, any dereference of the hash map pointer will access the wrong memory address (in this case an invalid memory address) and due to the corrupt incrementor value, the file contents are written to an arbitrary offset from the start of the buffer due to the corrupt incrementor value.  Because the code checks for a null pointer before cleaning up the hash map pointer, its value can either be replaced with null bytes or preserved as we did with the file pointer.  However, the incrementor variable must be set to a valid offset to continue storing the file’s contents and overwrite the return address.  If these addresses are written correctly, the exploit can overwrite the return address of the vulnerable function and gain control of execution.</p>

<p>By overflowing the buffer and writing specific values for three local variables, control of the return pointer is achieved, and <em>arbitrary code can be executed at the privilege level that <strong>bwa mem</strong> was run</em>.  The demo proof of concept exploit created uses a <code class="highlighter-rouge">ret2sys</code> style payload, passing a <code class="highlighter-rouge">sed</code> script as the argument to <code class="highlighter-rouge">system()</code> to perform a search and replace on a string sequence in all .FQ files in a directory.  The creation and display of this payload can be found in the figure below.  This payload was designed to highlight the issue that storage and processing of data, such as FASTQ files, frequently occur on the same system.  As a result, any amount of genomic data on this system could be corrupted.  However, the payload could just as easily have launched a shell or performed any operation allowed by the privilege level of the BWA process.</p>

<p><img src="assets/sedPayload.png" alt="sedPayload" /> <em>Payload that uses sed to search and replace a 17-character string in all .FQ files in the directory.</em></p>

<p>All in all, the vulnerability in BWA created an opportunity for arbitrary code execution, but the development of the exploit was made significantly simpler by relaxing several modern security protections. Writing a successful exploit for this vulnerability would be prohibitively difficult without local access to the system where BWA was running.  First, it would require trial and error to determine the location and values of the local variables to the vulnerable function.</p>

<p>Next, without ASLR being disabled, a Return-Oriented Programming (ROP) chain or some other method to nullify ASLR would have been necessary.  If stack smashing protection had been in place, the exploit would have needed to repair the stack canary.  Altogether, it is possible to write this exploit and launch it remotely, but it would have been significantly more difficult given a system and compiled application with these protections.</p>

<h2 id="creating-the-patch">Creating the patch</h2>

<p>The primary recommendation to developers on fixing the bug is simply to check the size of the .ALT file before reading its contents into memory.  Allocating a buffer large enough to hold the contents of the file will ensure that the buffer is not overrun and prevent arbitrary code execution.  Additionally, it is not standard security practice to use <code class="highlighter-rouge">fgetc()</code>, <code class="highlighter-rouge">strcat()</code>, or <code class="highlighter-rouge">strcpy()</code> because they do not verify the size of the destination buffer prior to execution. If these functions must be used, the size of the destination buffer should be checked before executing each function.  If these changes are implemented, this BWA vulnerability should be mitigated.  A proposed fix for the vulnerable code can be found in the figure below and has been distributed to the code maintainers.</p>

<p><img src="assets/bwaProposedCodeFix.png" alt="bwaProposedCodeFix" /> <em>Proposed patch.</em></p>

<h2 id="realistic-genomic-environment">Realistic genomic environment</h2>
<hr />

<p>The second major security assumption made in this assessment was that the index files were precompiled and delivered over insecure FTP. To evaluate the implications of this assumption, we created a model environment and attack to demonstrate a real-world scenario wherein the vulnerability could be exploited to modify patient data. This was created using our internal Emulytics<sup>TM</sup> capacity for simulation and assessment.<sup><a href="#ref6">6</a></sup> The model environment consisted of a network of three nodes resembling the client-server relationship that research facilities participate in to acquire reference genomes plus a malicious actor with presence on the research network.  In the demonstration a pipeline node equipped with BWA software requests reference genome files from a FTP server in the cloud, and another node existing on the pipeline node’s network is controlled by an attacker and launches an attack with the intent of corrupting patient data.</p>

<p><img src="assets/FTP_MITM_demo.png" alt="FTP_MITM_demo" /> <em><br /><br />MitM network demo. GRCh38 is the package of index files delivered by FTP. Blue folder is expected gzipped tarfile and red folder is the maliciously crafted gzipped tarfile.</em></p>

<p>We designed an attack scenario around this realistic network interaction involving a Man-in-the-Middle (MitM) of the FTP file transfer.  The attacker utilizes Bettercap, an extensible MitM framework written in Ruby, to intercept the files being transferred from the cloud to the pipeline and inject the malicious .ALT payload described in the Exploit Development section.  Because of the nuances of FTP, specifically in its usage of two separate connections for file transfer, we had to implement two Bettercap TCP proxy modules served from two MitM sessions.  The first session, control connection, needed to change the file size communicated to the client to reflect the injected data and prevent the connection from ending before the entire file was transferred.  Fortunately, because the injected data was smaller than the server’s original data in the demo, we did not encounter any TCP FIN packets from the server to modify or drop.  The second session, data connection, needed to simply replace the data being sent from the server with our malicious payload.</p>

<p>At the conclusion of the FTP session, the payload is delivered and ready to be launched.  Becuase of the legitimate server’s usage of FTP and our modifying traffic in transit, the victim is unaware of the introduction of the malicious payload and unknowingly launches the exploit when BWA-MEM is next run.</p>

<p><img src="assets/genomics_process.png" alt="genomics_process" /> <em>Head to Head description of full genomic pipeline and the impact of BWA indices with incorporated malware. BWA MEM does a search and replace of the sample files to corrupt the results. This results in changes that are propogated through the downstream programs.</em></p>

<h2 id="reasons-for-exploitability">Reasons for exploitability</h2>
<hr />

<ol>
  <li>BWA does not fail on poisoned .ALT file</li>
  <li>Easiest way to collect .ALT files is bwa.kit over FTP</li>
  <li>Files were not stored as read-only or in read-only directories</li>
  <li>Files did not have a database to protect them from change</li>
  <li>.ALT files were not checked after delivery</li>
</ol>

<h2 id="mitigation-recommendations">Mitigation Recommendations</h2>
<hr />

<p>First, our recommendation to mitigate the attack we demonstrated on BWA-MEM is to simply fix the vulnerability.  We exploited a classic buffer overflow vulnerability that can easily be resolved using more secure buffer allocation practices as described above and performed elsewhere in BWA’s code.  There is currently a patch existing for this bug, and we recommend that users download and install the most recent version of BWA <a href="https://github.com/lh3/bwa/tree/master">here</a>.</p>

<p>Our primary recommendation to mitigate attacks targeting other genomics software or other areas of the pipeline is to separate the storage of data with its processing.  Because these are two independent functions, they do not need to occur on the same system.  If the patient data (stored as FASTQ files) were not stored on the same system and not write-enabled, the exploit payload would not have been able to modify this data.  We believe this separation of responsibilities should be a standard best practice implemented by all facilities performing genomics research and processing.</p>

<p>Another best practice recommendation is to utilize file integrity checks, especially when downloading files over insecure protocols such as FTP or HTTP.  Integrity checks such as comparing file hashes using a modern cryptographic hashing algorithm (i.e. SHA-3) will reveal any modification to a .ALT or any other downloaded file as a result of a MitM.</p>
<p>MODIFICATION: As of 3/12/19 NCBI issued MD5 checksums on all of its files stored here. We encourage all users to use them and all softwares downloading data directly from <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/</a> to check the sums at <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/md5checksums.txt">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/md5checksums.txt</a> .</p>
    
<h2 id="summary-1">Summary</h2>
<hr />

<p>This study demonstrates a bug in the standard genomics pipeline. This bug was easily patched but points to a number of problems in standard genomics proceedure. We believe that fixing these issues will be valuable in making genomics, especially as it concerns patient medical genomics, safer and more secure.</p>

<h2 id="references">References</h2>
<hr />

<ul>
  <li><a name="ref1">[1]</a> <a href="http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1002195">Stephens, Zachary D., et al. “Big data: astronomical or genomical?.” PLoS biology 13.7 (2015): e1002195</a></li>
  <li><a name="ref2">[2]</a> <a href="https://academic.oup.com/database/article/doi/10.1093/database/bau069/2634692/OMICtools-an-informative-directory-for-multi-omic">Henry, Vincent J., et al. “OMICtools: an informative directory for multi-omic data analysis.” Database 2014 (2014)</a></li>
  <li><a name="ref3">[3]</a> <a href="http://www.internationalgenome.org/wiki/Analysis/vcf4.0/">VCF (Variant Call Format) version 4.0</a></li>
  <li><a name="ref4">[4]</a> <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4619817/pdf/BMRI2015-456479.pdf">Cornish, Adam, and Chittibabu Guda. “A comparison of variant calling pipelines using genome in a bottle as a reference.” BioMed research international 2015 (2015).</a></li>
  <li><a name="ref5">[5]</a> <a href="https://github.com/lh3/bwa/blob/master/README-alt.md">https://github.com/lh3/bwa/blob/master/README-alt.md</a></li>
  <li><a name="ref6">[6]</a> <a href="http://www.modsimworld.org/papers/2015/Emulytics_at_Sandia_National_Laboratories.pdf">Urias, V., et al. “Emulytics at Sandia National Laboratories.” MODSIM World (2015)</a></li>
</ul>

<hr />


  </div><a class="u-url" href="" hidden></a>
</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">Genomic Cybersecurity</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">Genomic Cybersecurity</li><li><a class="u-email" href="mailto:cmhudso@sandia.gov">cmhudso@sandia.gov</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/coreymhudson"><svg class="svg-icon"><use xlink:href="assets/minima-social-icons.svg#github"></use></svg> <span class="username">coreymhudson</span></a></li><li><a href="https://www.twitter.com/coreymhudson"><svg class="svg-icon"><use xlink:href="assets/minima-social-icons.svg#twitter"></use></svg> <span class="username">coreymhudson</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>This site describes cybersecurity research on genomic systems and platforms.</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
